File: app/src/main/java/org/dhis2/utils/filters/FiltersAdapter.java
package org.dhis2.utils.filters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.ObservableField;
import androidx.recyclerview.widget.RecyclerView;

import org.dhis2.data.tuples.Pair;
import org.dhis2.databinding.ItemFilterAssignedBinding;
import org.dhis2.databinding.ItemFilterCatOptCombBinding;
import org.dhis2.databinding.ItemFilterEnrollmentStatusBinding;
import org.dhis2.databinding.ItemFilterOrgUnitBinding;
import org.dhis2.databinding.ItemFilterPeriodBinding;
import org.dhis2.databinding.ItemFilterStateBinding;
import org.dhis2.databinding.ItemFilterStatusBinding;
import org.dhis2.utils.filters.sorting.SortingItem;
import org.hisp.dhis.android.core.category.CategoryCombo;
import org.hisp.dhis.android.core.category.CategoryOptionCombo;

import java.util.ArrayList;
import java.util.List;

public class FiltersAdapter extends RecyclerView.Adapter<FilterHolder> {

    private final ProgramType programType;
    private String enrollmentDateLabel;

    public enum ProgramType {ALL, EVENT, TRACKER, DATASET, DASHBOARD}

    private List<Filters> filtersList;
    private ObservableField<Filters> openedFilter;
    private ObservableField<SortingItem> sortingItem;
    private Pair<CategoryCombo, List<CategoryOptionCombo>> catCombData;

    public FiltersAdapter(ProgramType programType) {
        this.filtersList = new ArrayList<>();
        this.programType = programType;
        filtersList.add(Filters.PERIOD);
        filtersList.add(Filters.ORG_UNIT);
        filtersList.add(Filters.SYNC_STATE);
        openedFilter = new ObservableField<>();
        sortingItem = new ObservableField<>();
    }

    @NonNull
    @Override
    public FilterHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        switch (Filters.values()[viewType]) {
            case PERIOD:
                return new PeriodFilterHolder(ItemFilterPeriodBinding.inflate(inflater, parent, false), openedFilter, sortingItem, programType);
            case ENROLLMENT_DATE:
                return new EnrollmentDateFilterHolder(ItemFilterPeriodBinding.inflate(inflater, parent, false), openedFilter, sortingItem, programType);
            case ORG_UNIT:
                return new OrgUnitFilterHolder(ItemFilterOrgUnitBinding.inflate(inflater, parent, false), openedFilter, sortingItem, programType);
            case SYNC_STATE:
                return new SyncStateFilterHolder(ItemFilterStateBinding.inflate(inflater, parent, false), openedFilter, sortingItem, programType);
            case CAT_OPT_COMB:
                return new CatOptCombFilterHolder(ItemFilterCatOptCombBinding.inflate(inflater, parent, false), openedFilter, catCombData, programType);
            case EVENT_STATUS:
                return new StatusEventFilterHolder(ItemFilterStatusBinding.inflate(inflater, parent, false), openedFilter, programType);
            case ASSIGNED_TO_ME:
                return new AssignToMeFilterHolder(ItemFilterAssignedBinding.inflate(inflater, parent, false), openedFilter, programType);
            case ENROLLMENT_STATUS:
                return new StatusEnrollmentFilterHolder(ItemFilterEnrollmentStatusBinding.inflate(inflater, parent, false), openedFilter, sortingItem, programType);
            default:
                throw new IllegalArgumentException("Unsupported filter value");
        }
    }

    @Override
    public void onBindViewHolder(@NonNull FilterHolder holder, int position) {
        if (holder instanceof EnrollmentDateFilterHolder){
            ((EnrollmentDateFilterHolder)holder).updateLabel(enrollmentDateLabel).bind();
        } else {
            holder.bind();
        }
    }

    @Override
    public int getItemCount() {
        return filtersList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return filtersList.get(position).ordinal();
    }

    public void addCatOptCombFilter(Pair<CategoryCombo, List<CategoryOptionCombo>> categoryOptionCombos) {
        if (!filtersList.contains(Filters.CAT_OPT_COMB)) {
            filtersList.add(Filters.CAT_OPT_COMB);
            this.catCombData = categoryOptionCombos;
            notifyDataSetChanged();
        }
    }

    public void addEventStatus() {
        if (!filtersList.contains(Filters.EVENT_STATUS)) {
            filtersList.add(Filters.EVENT_STATUS);
            notifyDataSetChanged();
        }
    }

    public void addEnrollmentStatus() {
        if (!filtersList.contains(Filters.ENROLLMENT_STATUS)) {
            filtersList.add(Filters.ENROLLMENT_STATUS);
            notifyDataSetChanged();
        }
    }

    public void addAssignedToMe() {
        if (!filtersList.contains(Filters.ASSIGNED_TO_ME)) {
            filtersList.add(Filters.ASSIGNED_TO_ME);
            notifyDataSetChanged();
        }
    }

    public void addEnrollmentDate(String enrollmentDateLabel) {
        if (!filtersList.contains(Filters.ENROLLMENT_DATE)) {
            this.enrollmentDateLabel = enrollmentDateLabel;
            filtersList.add(0, Filters.ENROLLMENT_DATE);
            notifyDataSetChanged();
        }else if(enrollmentDateLabel!=null && !this.enrollmentDateLabel.equals(enrollmentDateLabel)){
            this.enrollmentDateLabel = enrollmentDateLabel;
            notifyDataSetChanged();
        }
    }

    public void removeAssignedToMe() {
        if (filtersList.contains(Filters.ASSIGNED_TO_ME)) {
            filtersList.remove(Filters.ASSIGNED_TO_ME);
        }
        FilterManager.getInstance().clearAssignToMe();
        notifyDataSetChanged();
    }

    public void removeEnrollmentDate() {
        if (filtersList.contains(Filters.ENROLLMENT_DATE)) {
            filtersList.remove(Filters.ENROLLMENT_DATE);
        }
        FilterManager.getInstance().clearEnrollmentDate();
        notifyDataSetChanged();
    }

}

File: app/src/main/java/org/dhis2/utils/filters/FiltersAdapter.kt
package org.dhis2.utils.filters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.ObservableField
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import org.dhis2.data.filter.FilterPresenter
import org.dhis2.databinding.ItemFilterAssignedBinding
import org.dhis2.databinding.ItemFilterCatOptCombBinding
import org.dhis2.databinding.ItemFilterEnrollmentStatusBinding
import org.dhis2.databinding.ItemFilterOrgUnitBinding
import org.dhis2.databinding.ItemFilterPeriodBinding
import org.dhis2.databinding.ItemFilterStateBinding
import org.dhis2.databinding.ItemFilterStatusBinding
import org.dhis2.databinding.ItemFilterWorkingListBinding
import org.dhis2.utils.filters.sorting.SortingItem

class FiltersAdapter(
    val programType: ProgramType,
    val filterPresenter: FilterPresenter
) : ListAdapter<FilterItem, FilterHolder>(object : DiffUtil.ItemCallback<FilterItem>() {
    override fun areItemsTheSame(oldItem: FilterItem, newItem: FilterItem): Boolean {
        return oldItem == newItem
    }

    override fun areContentsTheSame(oldItem: FilterItem, newItem: FilterItem): Boolean {
        return oldItem == newItem
    }
}) {
    private val openedFilter: ObservableField<Filters> = ObservableField()
    private val sortingItem: ObservableField<SortingItem> = ObservableField()

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): FilterHolder {
        val inflater =
            LayoutInflater.from(parent.context)
        return when (Filters.values()[viewType]) {
            Filters.PERIOD -> PeriodFilterHolder(
                ItemFilterPeriodBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, sortingItem, programType
            )
            Filters.ENROLLMENT_DATE -> EnrollmentDateFilterHolder(
                ItemFilterPeriodBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, sortingItem, programType
            )
            Filters.ORG_UNIT -> OrgUnitFilterHolder(
                ItemFilterOrgUnitBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, sortingItem, programType, filterPresenter
            )
            Filters.SYNC_STATE -> SyncStateFilterHolder(
                ItemFilterStateBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, sortingItem, programType
            )
            Filters.CAT_OPT_COMB -> CatOptCombFilterHolder(
                ItemFilterCatOptCombBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, null, programType
            )
            Filters.EVENT_STATUS -> StatusEventFilterHolder(
                ItemFilterStatusBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, programType
            )
            Filters.ASSIGNED_TO_ME -> AssignToMeFilterHolder(
                ItemFilterAssignedBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, programType
            )
            Filters.ENROLLMENT_STATUS -> StatusEnrollmentFilterHolder(
                ItemFilterEnrollmentStatusBinding.inflate(
                    inflater,
                    parent,
                    false
                ),
                openedFilter,
                sortingItem,
                programType
            )
            Filters.WORKING_LIST -> WorkingListFilterHolder(
                ItemFilterWorkingListBinding.inflate(
                    inflater,
                    parent,
                    false
                ), openedFilter, programType
            )
        }
    }

    override fun onBindViewHolder(holder: FilterHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun getItemViewType(position: Int): Int {
        return getItem(position).type.ordinal
    }
}